### PR TITLE
Instructions for connecting VS Code remotely

### PR DESCRIPTION
**Pull request by Jongmassey**
- SSH Server setup
- VS Code Remote SSH extension installation and configuration
- Gitpod-specific configuration
- Supporting images

### CODE REVIEW

**Comment by StevenMaude** [comment]
@Jongmassey: thanks! I can take a look at this next week if no-one else does in-between; give me a nudge should I forget.

**Comment by Jongmassey** [comment]
I've just noticed I've not added this to the nav menu as I wasn't sure where it should go

**Comment by StevenMaude** [comment]
> I've just noticed I've not added this to the nav menu as I wasn't sure where it should go

As the site is structured now, probably "Reference > Detailed installation guides" fits best, to my eye :eyes:

File: docs/working-remotely-ssh.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Working remotely with Visual Studio Code
+
+It is possible to work remotely on a machine containing the required opensafely software using Visual Studio Code.
```
Comment (StevenMaude) [inline docs/working-remotely-ssh.md:3]: nit: "opensafely software" here should probably be "OpenSAFELY software".

(Aside: Across the documentation, I'm not sure that we have a strong convention for distinguishing "all the software you need to run OpenSAFELY" from "the OpenSAFELY command-line tool: `opensafely`" — which is also software and a subset of the first.)

File: docs/working-remotely-ssh.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+This may be useful in cases where it is not possible to install all of the required sofware to run opensafely on your local machine, or where running this required software severly affects the performance of your local machine.
```
Comment (StevenMaude) [inline docs/working-remotely-ssh.md:4]: nit: "severly" should be "severely"

File: docs/working-remotely-ssh.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+This may be useful in cases where it is not possible to install all of the required sofware to run opensafely on your local machine, or where running this required software severly affects the performance of your local machine. 
+
+This guide walks you through the steps required to set up the machine running the opensafely software (the "server") and the local machine from which you will access it (the "client").
+
+## 1. Server setup
+
+First, ensure that the opensafely software and all of its pre-requisites have been installed using the [Getting Started guide](getting-started). If using this guide to connect to Gitpod, please skip to step 2.
+
+### SSH server installation and configuration
+The steps in this guide require enabling access to the server using secure shell protocol (SSH).
```
Comment (StevenMaude) [inline docs/working-remotely-ssh.md:13]: nit: I think "secure shell protocol" is conventionally capitalised as "Secure Shell Protocol", e.g [see Wikipedia's page](https://en.wikipedia.org/wiki/Secure_Shell_Protocol) and the RFC documents linked from there.

File: docs/working-remotely-ssh.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+ ![Visual Studio Code Remote Extension](images/vs-code-remote-extension.png "Visual Studio Code Remote Extension")
+
+Once installed, connect to the server by opening the Command Palette (Ctrl+Shift+P, Cmd+Shift+P on Mac) and start typing "Remote-SSH" then select `Remote-SSH:Connect to Host...`.
+
+![Visual Studio Code Connect to Remote SSH Host](images/vs-code-command-palette-remote-ssh-connect.png "Visual Studio Code Connect to Remote SSH Host")
+
+
+Visual Studio Code will then guide you through logging in to the server and will automatically install its own requirements on the server.
+Once connected, you may open a folder on the server within Visual Studio code, edit files, and issue commands to the terminal as if working on your local machine.
```
Comment (StevenMaude) [inline docs/working-remotely-ssh.md:36]: nit: "Visual Studio code" should be "Visual Studio Code"

File: docs/working-remotely-ssh.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+On Windows, this requires installation of the [OpenSSH Server optional feature](https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_install_firstuse).
+On Linux, this requires installation of the OpenSSH Server such as is described in these [instructions for Ubuntu Linux](https://ubuntu.com/server/docs/service-openssh).
+
+### Git configuration
+This guide assumes you have set up a GitHub user account as per the [Getting Started guide](getting-started).  
+If you have not already set up git and logged into GitHub from the server it may be neccesary to [configure your identity](https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup#_your_identity).
+If problems are encountered with logging into github when working remotely, it may be helpful to make use of the [git credential caching](https://docs.github.com/en/get-started/getting-started-with-git/caching-your-github-credentials-in-git) feature, or to [connect to GitHub using SSH](https://docs.github.com/en/github/authenticating-to-github/connecting-to-github-with-ssh).
```
Comment (StevenMaude) [inline docs/working-remotely-ssh.md:21]: nit: I'd capitalise the "github" to "GitHub" for consistency with the other uses here.

File: docs/working-remotely-ssh.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+![Visual Studio Code Connect to Remote SSH Host](images/vs-code-command-palette-remote-ssh-connect.png "Visual Studio Code Connect to Remote SSH Host")
+
+
+Visual Studio Code will then guide you through logging in to the server and will automatically install its own requirements on the server.
+Once connected, you may open a folder on the server within Visual Studio code, edit files, and issue commands to the terminal as if working on your local machine. 
+
+## 3. (Optional) Connecting to GitPod
+
+Using Visual Studio Code to connect to Gitpod over SSH is possible via [Gitpod's Local Companion app](https://www.gitpod.io/blog/local-app#installation). Once installed, and authenticated to your Gitpod account, the local companion app will create a temporary SSH configuration file with an entry for each one of your workspaces. On Linux and MacOS, this file is located at `/tmp/gitpod_ssh_config` and on Windows in `%TEMP%\gitpod_ssh_config`.
```
Comment (StevenMaude) [inline docs/working-remotely-ssh.md:40]: nit: MacOS should be macOS, in line with Apple's usage, and usage in this document above.

Comment (StevenMaude) [reply docs/working-remotely-ssh.md:40]: I'd consider saying "each one of your **Gitpod** workspaces" (emphasis for clarity only here).

Workspace is a terribly overloaded term: Gitpod has workspaces, VSCode has workspaces, OpenSAFELY has workspaces…

File: docs/working-remotely-ssh.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
+# Working remotely with Visual Studio Code
+
+It is possible to work remotely on a machine containing the required opensafely software using Visual Studio Code.
+This may be useful in cases where it is not possible to install all of the required sofware to run opensafely on your local machine, or where running this required software severly affects the performance of your local machine. 
+
+This guide walks you through the steps required to set up the machine running the opensafely software (the "server") and the local machine from which you will access it (the "client").
+
+## 1. Server setup
+
+First, ensure that the opensafely software and all of its pre-requisites have been installed using the [Getting Started guide](getting-started). If using this guide to connect to Gitpod, please skip to step 2.
```
Comment (StevenMaude) [inline docs/working-remotely-ssh.md:10]: nit: you can drop the hyphen from "pre-requisites".

File: docs/working-remotely-ssh.md (hunk 2 above)
Comment (StevenMaude) [reply docs/working-remotely-ssh.md:4]: Maybe worth clarifying that there is a requirement that the server requires that its network connection allows connections on a port open for SSH, and that the client machine is able to make connections to that port? Some users might be on networks with a firewall that they do not control.

Also maybe worth emphasising that you should have permission from whoever is running the server to do so, if that is not you? In principle, you would require admin access to configure this, but still.

File: docs/working-remotely-ssh.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+### SSH server installation and configuration
+The steps in this guide require enabling access to the server using secure shell protocol (SSH).
+On macOS, this requires enabling the [Remote Login](https://support.apple.com/en-gb/guide/mac-help/mchlp1066/mac) feature. 
+On Windows, this requires installation of the [OpenSSH Server optional feature](https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_install_firstuse).
+On Linux, this requires installation of the OpenSSH Server such as is described in these [instructions for Ubuntu Linux](https://ubuntu.com/server/docs/service-openssh).
+
```
Comment (StevenMaude) [inline docs/working-remotely-ssh.md:17]: If we're having to explain at a basic level how to configure SSH, maybe we should add a mention that it is better to authenticate with a public/private key, over a password, and to set a passphrase for that private key.

* I'm not entirely sure if the default for Windows is to allow username/password access which isn't very good; see this [key management page from Microsoft](https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_keymanagement for how to do so). From a quick look, I also think the macOS default is to allow password authentication.
* The linked Ubuntu guide does explain how to generate a key pair. However, it doesn't make the importance of setting a passphrase for a private key clear. Those instructions state "During the process you will be prompted for a password. Simply hit *Enter* when prompted to create the key."

**Comment by StevenMaude** [comment]
Despite my nit picking, this looks good to me and could be helpful to more technically inclined OpenSAFELY users who want to work around restrictions they're facing.

I added a couple of suggestions regarding small possible tweaks for clarification, given the possible audience, and where the document should go.

Finally, for good cleanliness and our future selves, we could tidy the commit history here by "squashing" the multiple commits into just one single commit, e.g. "Add instructions for connecting to VS Code remotely". [1] But that's also less essential here than it would be in a repository containing code.

[1] At the command line, you can do this locally (`git checkout vs-code-remote-ssh`, then `git rebase -i` and a `git push --force-with-lease`). Apparently, you can also [squash](https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges) when merging the pull request: click the▼ to see the option to do so, next to the "Merge pull request" button.

**Comment by StevenMaude** [comment]
So, I'll approve in current state and let @Jongmassey decide what he wants to do.

**Review by StevenMaude — APPROVED** [review]

**Comment by Jongmassey** [comment]
Nits have been picked, I'll have a think about the more substantial comments at a more civilised hour

**Comment by HelenCEBM** [comment]
@Jongmassey any further thoughts on this?

**Comment by Jongmassey** [comment]
I think at the time I wasn't sure what the protocol for merging was. If the approval still stands @StevenMaude  then I'll merge

**Comment by StevenMaude** [comment]
@Jongmassey: sorry about being unresponsive here. I missed this with being away and then trying to catch up after being away.

You might want to find a suitable place in the document hierarchy to add it in `mkdocs.yml`. I think, right now, it only shows up if you search for something in the content.

Probably needs a rebase this branch on `main` first because `mkdocs.yml` changed considerably after @HelenCEBM did a tidy up in #392.

When that's done, it's fine to merge :tada:

**Comment by Jongmassey** [comment]
I think  this is largely irrelevant now, especially with native gitpod VS code connection extension

**Comment by HelenCEBM** [comment]
Any bits to keep in or shall we close the PR?

**Comment by StevenMaude** [comment]
@Jongmassey: if the [Gitpod Visual Studio Code integration](https://www.gitpod.io/docs/editors/vscode) solves this problem[^1], then maybe just replace most of this with a mention of that in `docs/gitpod-workspaces.md`?

If the reader has done some coding before, they might welcome using a native local editor, even if they can't install Docker and the rest of the software.

[^1]: VS Code works [without needing Windows admin permissions](https://code.visualstudio.com/docs/editor/portable), as far as I can tell.

**Comment by CarolineMorton** [comment]
Just having a look at our PRs, what do you want to do with this @Jongmassey ?

**Pull request closed without merging**